core/metrics-spi/src/main/java/software/amazon/awssdk/metrics/provider/SystemSettingsMetricConfigurationProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.provider;
+
+import java.util.HashSet;
+import java.util.Set;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.metrics.MetricCategory;
+import software.amazon.awssdk.metrics.internal.MetricSystemSetting;
+
+/**
+ * A {@link MetricConfigurationProvider} implementation that uses system properties to configure the metrics feature.
+ * <p>
+ * As Java system properties can only be set during application start time or using application code,
+ * it is not possible to change the behavior at runtime without code changes.
+ */
+@SdkProtectedApi
+public final class SystemSettingsMetricConfigurationProvider implements MetricConfigurationProvider {
+
+    private final boolean enabled;
+    private final Set<MetricCategory> categories;
+
+    private SystemSettingsMetricConfigurationProvider() {
+        this.enabled = resolveEnabled();
+        this.categories = resolveCategories();
+    }
+
+    public static SystemSettingsMetricConfigurationProvider create() {
+        return new SystemSettingsMetricConfigurationProvider();
+    }
+
+    @Override
+    public boolean enabled() {
+        return enabled;
+    }
+
+    @Override
+    public Set<MetricCategory> metricCategories() {
+        return categories;
```

Comment:
should return immutable set

Reply:
In this case, can switch the resolveCategories to use an immutable set. In the future, it's possible to change these at runtime so this would introduce a bunch of copying as this method is called frequently...

Reply:
:+1: